### PR TITLE
Conservation Efforts Widget

### PR DESCRIPTION
**Pull request by aabdaab**
This is a first iteration of [Conservation Efforts Widget](https://projects.invisionapp.com/d/main?origin=v7?origin=v7#/console/17484384/365514646/inspect) in the landscape sidebar.

- Added custom pie chart which `explodes` when certain layers are activated / already rendered on the map (protected and community-based layers)
- Fed raw data to the chart from all aggregated cells in the viewport (same logic as for human pressures widget)
- Data is taken from this arcgis [table](https://eowilson.maps.arcgis.com/home/item.html?id=74918f5538524c9f8e51554df8b0ac81#data)
The algorithm for painting specific slices is as follows:
```
if WDPA_prop + RAISG_prop > all_prop:
               yellow = RAISG_prop
               orange = all_prop - RAISG_prop
               grey = 1 - (orange + RAISG_prop)
else 
               yellow = RAISG_prop
               orange = WDPA_prop
               grey = 1 - (RAISG_prop + WDPA_prop)
```
I think the algorithm needs to be improved and possibly a pie chart in this shape is not the best solution (it seems to me that community-based areas need to be included IN the wdpa slice as it is on the map).

When talking to @gretacv I suggested this approach (should be the other way around than what we see on the image - community-based should be the smaller part of all protected areas):
Image from iOS

Example of calculated data on the map:

image

### CODE REVIEW

**Comment by gretacv** [comment]
There is a problem with the sentence.

**Comment by aabdaab** [comment]
Ready to review ☝️The GEE has completed so you can see the data after the table was updated.

image
```
if WDPA_prop + RAISG_prop > all_prop: 
  yellow = all_prop - WDPA_prop 
  orange = WDPA_prop 
  grey = 1 - (orange + RAISG_prop) 
else 
  yellow = RAISG_prop 
  orange = WDPA_prop 
  grey = 1 - (RAISG_prop + WDPA_prop)
```
@gretacv please take a look, still something's wrong. Either, the sentence is wrong (should be 70,75% as we can see on the pie chart) or the pie chart itself is wrong (yellow should be within the red area). Let me know your thoughts 🙌

**Comment by kfilip94** [comment]
checking :eyes:

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-styles.module.scss [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    bottom: $margin;
+  }
+}
+
+.padding {
+  padding: 20px;
+}
+
+.title {
+  margin: 0;
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-styles.module.scss]: Please import `@extend %title;` from `typography-extends` file

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-styles.module.scss [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    height: 10px;
+    position: absolute;
+    right: 0px;
+    top: 6px;
+    width: 10px;
+  }
+}
+
+.description {
+  color: $alto;
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-styles.module.scss]: Here you can import `@extend %annotation;` from `typography-extends` file.

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-component.jsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+          url: "https://services9.arcgis.com/IkktFdUAcY3WrH25/arcgis/rest/services/ConsProp/FeatureServer"
+        });
+        setConservationPropsLayer(consPropLayer)
+      });
+    }
+  }, []);
+
+  const queryParams = conservationPropsLayer && conservationPropsLayer.createQuery();
+
+  const alreadyChecked = WDPALayers.reduce((acc, option) => ({
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-component.jsx]: Maybe would be worth to move it to selectors?

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-component.jsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+    handleLayerToggle(option.id);
+    handleLayerRendered(view, layerToggled, handleGlobeUpdating);
+
+    const isLayerActive = alreadyChecked[option.value];
+    if (isLayerActive) addLayerAnalyticsEvent({ slug: option.slug })
+    else removeLayerAnalyticsEvent({ slug: option.slug });
+  }
+
+  const activeSlices = rawData && Object.keys(rawData)
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-component.jsx]: [2] Maybe would be worth to move it to selectors?

**Comment by aabdaab** [comment]
Updated description text, now it's correct 👍 
image

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-component.jsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    if (terrestrialCellData && queryParams) {
+      queryParams.where = `CELL_ID IN (${terrestrialCellData.map(i => i.CELL_ID).join(', ')})`;
+      conservationPropsLayer.queryFeatures(queryParams).then(function(results){
+        const { features } = results;
+        setConservationEfforts(features.map(c => c.attributes));
+      });
+    }
+  }, [terrestrialCellData])
+
+  const toggleLayer = (layersPassed, option) => {
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-component.jsx]: `toggleLayer` function you can move to `conservation-efforts-widget` container

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-selectors.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  }
+)
+
+const getConservationAreasLogic = createSelector(
+  [getConservationEfforts],
+  (conservationEfforts) => {
+    if (!conservationEfforts) return null;
+
+    const areas = {};
+    if (conservationEfforts.WDPA_prop + conservationEfforts.RAISG_prop > conservationEfforts.all_prop) {
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-selectors.js]: Here I'd recommend using destructuring, just to increase readability of this condition :) :
`const { WDPA_prop, RAISG_prop, all_prop } = conservationEfforts;`

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-selectors.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    return areas;
+  }
+)
+
+const getConservationAreasFormatted = createSelector(
+  [getConservationAreasLogic],
+  (conservationAreasLogic) => {
+    if (!conservationAreasLogic) return null;
+
+    const formattedAreas = Object.values(conservationAreasLogic).reduce((obj, key) => {
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-selectors.js]: I think that you can do it simpler here:

``
```suggestion
    const formattedAreas = Object.keys(conservationAreasLogic).reduce((acc, key) => {
      return { ...acc, [key]: conservationAreasLogic[key].toFixed(2) };
    }, {});
```

File: src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-styles.module.scss [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  @extend  %bodyText;
+  color: $white;
+  padding: 0 20px;
+  position: relative;
+  margin: 8px 0 0;
+  width: max-content;
+
+  &:after {
+    content: '';
+    background: #6C828F;
```
Comment (kfilip94) [inline src/components/landscape-sidebar/conservation-efforts-widget/conservation-efforts-widget-styles.module.scss]: We have this color already, use `$slate-gray` here :)

File: src/components/pie-chart/pie-chart-component.jsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        const strokeWidth = active ? EXPLODING_SLICE_STROKE_WIDTH : '';
+
+        return (
+          <Slice 
+            key={i}
+            outerRadius={radius}
+            stroke={stroke}
+            strokeWidth={strokeWidth}
+            value={value}
+            fill={colors[getKeyByValue(filteredChartData, value.data)]}
```
Comment (kfilip94) [inline src/components/pie-chart/pie-chart-component.jsx]: Use area here, you have it already
```suggestion
            fill={colors[area]}
```

File: src/components/pie-chart/pie-chart-component.jsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  const filteredChartData = chartData && Object.keys(chartData)
+    .filter((key) => chartData[key])
+    .reduce((obj, key) => {
+      obj[key] = chartData[key];
+      return obj;
+    }, {});
+
+  return (
+    <g id="conservation-widget" transform={translate(x, y)}>
+      {filteredChartData && pie(Object.values(filteredChartData)).map((value, i) => {
```
Comment (kfilip94) [inline src/components/pie-chart/pie-chart-component.jsx]: You can turn this bit and `filteredChartData` into selector function

**Comment by kfilip94** [comment]
@aabdaab Good job with this widget!
I found a few small bugs here:
 -  We can hide this widget when the data is still loading:
3dngn-fbpzj

- After refreshing the page, there is a small bug with slices order in pie chart:
Screenshot from 2019-09-11 15 36 41
- Pie chart itself should be bigger, according to the design it should be 120px x 120px,  but now it's 100px

**Comment by kfilip94** [comment]
I will take care of these changes :point_up: (approved by @aabdaab :heavy_check_mark:)

**Comment by kfilip94** [comment]
@aabdaab @weberjavi  :point_up: :point_up_2: :arrow_up: :arrow_up_small: :up: ready again!  :tada:

**Comment by aabdaab** [comment]
👏

**Pull request merged**